20220714-lookup-api.md

```diff
@@ -0,0 +1,462 @@
+
+The `/stores/{store_id}/lookup:streamed` HTTP path mapping for the streaming variant was inspired by [Google's HTTP mapping for Custom Methods](https://cloud.google.com/apis/design/custom_methods), but this is open for debate.
+
+If anyone has any opinions or recommendations we'd love to hear them!
+
+## Configuration Changes
+- Add a `StreamedLookupDeadline` config param (`--streamed-lookup-deadline` flag)
+
+  This will be used to define the maximum amount of time to stream Lookup results until the server stream is closed. This is to protect the server from misuse of the streaming endpoint. If this config is omitted all object ids will be streamed back to the client (regardless of how many there are).
+
+- Add a `LookupMaxResults` config param (`--lookup-max-results` flag)
+
+  This will be used to define the maximum list size of the list of object ids that are returned in the unary Lookup API endpoint. If this is omitted all object ids will be fetched (regardless of how many there are).
+  
+For production usage we should document recommendations for these configuration limits so as to protect the server from memory exhaustion or costly Lookup queries.
+
+## Server Changes
+* Add implementations of the API changes described in [Changes to the OpenFGA API](#changes-to-the-openfga-api) in the server. This includes implementations:
+```
+func (s *Server) StreamedLookup(
+    req *openfgapb.LookupRequest,
+    srv openfgapb.OpenFGAService_StreamedLookupServer,
+) error {...}
+
+func (s *Server) Lookup(
+    ctx context.Context,
+    req *openfgapb.LookupRequest,
+) (*openfgapb.LookupResponse, error) {...}
+```
+
+Sample code implementing these proposed changes can be found here:
+https://github.com/openfga/openfga/tree/lookup-poc (server changes) which depends on https://github.com/openfga/api/tree/lookup-api (protobuf API changes)
+
+### Storage Interface Changes
+The [`OpenFGADatastore`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L132) interface will need to be expanded or modified to support queries for all objects of a particular object type (within a store). Today our datastore/storage interface only has the [`Read`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L51) method, but it requires a storeID and tuple key specifying either the `object` or `user` field (or both). The Lookup work will need a method that supports only providing the storeID and object `type`.
+
+We should modify the [`TupleBackend`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L47) interface with a new method `QueryRelationshipTuples` and it's signature could look like:
```

Comment:
The [`TupleBackend`](https://github.com/openfga/openfga/blob/170a6834d057428e3b0d250cae47a01f5a61898f/storage/storage.go#L47) interface will need a new method `QueryRelationshipTuples` and it's signature could look like: